neural_lam/weather_dataset.py

```diff
@@ -1,262 +1,373 @@
+        # non-zero amount of samples
+        if self.__len__() <= 0:
+            raise ValueError(
+                "The provided datastore only provides "
+                f"{len(self.da_state.time)} time steps for `{split}` split, "
+                f"which is less than the required 2+ar_steps "
+                f"(2+{self.ar_steps}={2 + self.ar_steps}) for creating a "
+                "sample with initial and target states."
+            )
 
         # Set up for standardization
+        # TODO: This will become part of ar_model.py soon!
         self.standardize = standardize
         if standardize:
-            ds_stats = utils.load_dataset_stats(dataset_name, "cpu")
-            self.data_mean, self.data_std, self.flux_mean, self.flux_std = (
-                ds_stats["data_mean"],
-                ds_stats["data_std"],
-                ds_stats["flux_mean"],
-                ds_stats["flux_std"],
+            self.ds_state_stats = self.datastore.get_normalization_dataarray(
+                category="state"
             )
 
-        # If subsample index should be sampled (only duing training)
-        self.random_subsample = split == "train"
+            self.da_state_mean = self.ds_state_stats.state_mean
+            self.da_state_std = self.ds_state_stats.state_std
+
+            if self.da_forcing is not None:
+                self.ds_forcing_stats = (
+                    self.datastore.get_normalization_dataarray(
+                        category="forcing"
+                    )
+                )
+                self.da_forcing_mean = self.ds_forcing_stats.forcing_mean
+                self.da_forcing_std = self.ds_forcing_stats.forcing_std
 
     def __len__(self):
-        return len(self.sample_names)
+        if self.datastore.is_forecast:
+            # for now we simply create a single sample for each analysis time
+            # and then the next ar_steps forecast times
+            if self.datastore.is_ensemble:
+                warnings.warn(
+                    "only using first ensemble member, so dataset size is "
+                    " effectively reduced by the number of ensemble members "
+                    f"({self.da_state.ensemble_member.size})",
+                    UserWarning,
+                )
+            # XXX: we should maybe check that the 2+ar_steps actually fits in
+            # the elapsed_forecast_duration dimension, should that be checked
+            # here?
+            return self.da_state.analysis_time.size
+        else:
+            # sample_len = 2 + ar_steps
+            #             (2 initial states + ar_steps target states)
+            # n_samples = len(self.da_state.time) - sample_len + 1
+            #           = len(self.da_state.time) - 2 - ar_steps + 1
+            #           = len(self.da_state.time) - ar_steps - 1
+            return len(self.da_state.time) - self.ar_steps - 1
+
+    def _sample_time(self, da, idx, n_steps: int, n_timesteps_offset: int = 0):
```

Comment:
I was very confused by this method name, as this method does not perform any sampling. Propose to change to something like `_time_slice_da`.